File: 393-utf-8-validation/validUtf8.cpp
#include <iostream>
#include <vector>

using namespace std;
/**
    这题不是很难，但是需要理解清楚题意，
    一开始没加 if (bytes > 4) return false;
    提交了两次都没通过
*/
bool validUtf8(vector<int>& data) {
    int n = data.size();
    if (n == 0) return true;

    int bytes = 0;
    for (int i = 0; i < n; i++) {
        int first = data[i];
        if (bytes > 0) {
            if ((first & 192) != 128) {
                return false;
            }
            bytes--;
        }
        // 1xxxxxxx
        else if (first & 128) {
            // 不允许出现 10xxxxxx
            if (!(first & 64)) return false;
            while ((first & 128)) {
                bytes++;
                first <<= 1;
            }
            if (bytes > 4) return false;
            bytes--;
        }
    }
    return bytes == 0;
}
int main() {
    vector<int> data = {197, 130, 1};
    cout << validUtf8(data) << endl;
    return 0;
}

File: 393-utf-8-validation/validUtf8.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

class Solution(object):
    def validUtf8(self, data):
        """
        :type data: List[int]
        :rtype: bool
        """
        n = len(data)
        if n == 0: return True

        bytes = 0
        for i, val in enumerate(data):
            if bytes > 0:
                if val & 192 != 128: return False
                bytes -= 1
            elif val & 128 != 0:
                if val & 64 == 0: return False
                while val & 128 != 0:
                    bytes += 1
                    val <<= 1
                if bytes > 4: return False
                bytes -= 1
        return bytes == 0

s = Solution()
print(s.validUtf8([235, 140, 4]))
